proposals/lip-introduce_bft_module.md

```diff
@@ -0,0 +1,805 @@
+
+##### Returns
+
+For a height `h` as input, the function returns the smallest height `h1 > h` with an entry in the BFT Parameters substore with key `h1`, if such height exists. Otherwise, it fails.
+
+
+#### setBFTParameters
+
+This function allows to set the thresholds, validators and associated BFT weights to be used from the next height onward and updates the BFT module store accordingly.
+
+##### Parameters
+
+* `precommitThreshold`: The precommit threshold value to be used from the next height onward. The value must be a 64-bit unsigned integer.
+* `certificateThreshold`: The certificate threshold value to be used from the next height onward. The value must be a 64-bit unsigned integer.
+* `validators`: The validators and their associated BFT weight to be used from the next height onward. The value must be an array of objects with an `address` property containing the 20-byte address of the validator and a `bftWeight` property containing the BFT weight of the validator as 64-bit unsigned positive integer.
+Note that this function fails if the input contains validators with zero BFT weight or if the number of provided validators is more than `LSK_BFT_BATCH_SIZE`.
+
+##### Execution
+
+```python
+setBFTParameters(precommitThreshold, certificateThreshold, validators):
+    if length(validators) > LSK_BFT_BATCH_SIZE:
+        fail
+    aggregateBFTWeight = 0
+    for validator in validators:
+        if validator.bftWeight <= 0:
+            fail
+        aggregateBFTWeight += validator.bftWeight
+    if floor(1/3*aggregateBFTWeight)+1 > precommitThreshold or precommitThreshold > aggregateBFTWeight:
+        fail
+    if floor(1/3*aggregateBFTWeight)+1 > certificateThreshold or certificateThreshold > aggregateBFTWeight:
+        fail
+    bftParams = object following bftParametersSchema
+    bftParams.prevoteThreshold = floor(2/3*aggregateBFTWeight)+1
+    bftParams.precommitThreshold = precommitThreshold
+    bftParams.certificateThreshold = certificateThreshold
+    sort validators lexicographically by address
+    bftParams.validators = validators
+
+    # compute the validators hash
+    activeValidators = []
+    for validator in validators:
+        activeValidator = object with property bftWeight and blsKey
+        activeValidator.bftWeight = validator.bftWeight
+        activeValidator.blsKey =  validatorModule.getValidatorAccount(validator.address).blsKey
+        add activeValidator to activeValidators
+    sort elements in activeValidators lexicographically by blsKey property
+    bftParams.validatorsHash = computeValidatorsHashInternal(activeValidators, certificateThreshold)
+
+    # set the BFT parameters for the correct next height
+    if length(bftVotes.blockBFTInfos) > 0:
+        nextHeight = bftVotes.blockBFTInfos[0].height + 1
+    else:
+        # only directly after processing the genesis block we have
+        # length(bftVotes.blockBFTInfos) == 0
+        nextHeight = bftVotes.maxHeightPrevoted + 1
```

Comment:
when this is called during genesis block creation or execution, (for example h = 0), it will set this `nextHeight` to `1`. 
to get the `validatorsHash` for the genesis block, it will call `getBFTParameters` with `h = 0` and it will fail because there is no parameter for `h = 0`

Reply:
The problem was when calling `getBFTParameters`, it needed to be called with h + 1 as discussed. There is no issue on the specification.